File: android/app/src/main/java/cz/arnal/bleedge/core/Fragment.kt
package cz.arnal.bleedge.core

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.concurrent.ConcurrentHashMap
import java.util.zip.CRC32
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

// Frame wire encoding: version(1) + packet_id(16) + frag_idx(1) + frag_count(1) + crc32(4) + data
private const val FRAME_HEADER_SIZE = 23

// MAX_FRAME_SIZE is the maximum GATT frame (single ATT write) size used for
// fragmentation. It must stay <= the smallest peer's negotiated (ATT_MTU - 3);
// the ESP32 relay negotiates an ATT MTU of 247, so 200 is safe for every peer.
// Mirrors core.MaxFrameSize (Go) and FRAGMENT_MTU (firmware). See PROTOCOL.md.
const val MAX_FRAME_SIZE = 200

/**
 * Frame is the GATT transport unit. One or more frames reassemble into a Packet's encoded bytes.
 */
data class Frame(
    val version: Byte,
    val packetId: ByteArray,    // 16 bytes
    val fragmentIndex: Byte,
    val fragmentCount: Byte,
    val payloadCrc32: Int,      // unsigned CRC32 stored as Int
    val data: ByteArray,
) {
    fun encode(): ByteArray {
        val buf = ByteBuffer.allocate(FRAME_HEADER_SIZE + data.size).apply {
            order(ByteOrder.BIG_ENDIAN)
            put(version)
            put(packetId)
            put(fragmentIndex)
            put(fragmentCount)
            putInt(payloadCrc32)
            put(data)
        }
        return buf.array()
    }

    companion object {
        fun decode(raw: ByteArray): Frame {
            require(raw.size >= FRAME_HEADER_SIZE) {
                "Frame too short: ${raw.size} bytes (need $FRAME_HEADER_SIZE)"
            }
            val buf = ByteBuffer.wrap(raw).order(ByteOrder.BIG_ENDIAN)
            val version = buf.get()
            val packetId = ByteArray(16).also { buf.get(it) }
            val fragmentIndex = buf.get()
            val fragmentCount = buf.get()
            val crc32 = buf.int
            val data = ByteArray(raw.size - FRAME_HEADER_SIZE).also { buf.get(it) }
            return Frame(version, packetId, fragmentIndex, fragmentCount, crc32, data)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Frame) return false
        return version == other.version &&
            packetId.contentEquals(other.packetId) &&
            fragmentIndex == other.fragmentIndex &&
            fragmentCount == other.fragmentCount &&
            payloadCrc32 == other.payloadCrc32 &&
            data.contentEquals(other.data)
    }

    override fun hashCode(): Int = packetId.contentHashCode() * 31 + fragmentIndex
}

/**
 * Splits [packetData] into GATT frames of at most [mtu] bytes total.
 */
fun fragmentPacket(packetData: ByteArray, mtu: Int, packetId: ByteArray): List<Frame> {
    val maxData = (mtu - FRAME_HEADER_SIZE).coerceAtLeast(1)
    val crc = CRC32().also { it.update(packetData) }.value.toInt()

    val frames = mutableListOf<Frame>()
    var offset = 0
    while (offset < packetData.size) {
        val end = (offset + maxData).coerceAtMost(packetData.size)
        frames += Frame(
            version = 1,
            packetId = packetId,
            fragmentIndex = frames.size.toByte(),
            fragmentCount = 0, // filled below
            payloadCrc32 = crc,
            data = packetData.copyOfRange(offset, end),
        )
        offset = end
    }
    val count = frames.size.toByte()
    return frames.map { it.copy(fragmentCount = count) }
}

private data class Assembly(
    val frags: MutableMap<Int, ByteArray> = mutableMapOf(),
    val count: Int,
    val crc: Int,
    var lastSeen: Long = System.currentTimeMillis(),
)

/**
 * Reassembler collects frames and reassembles complete packets.
 * Thread-safe. Background coroutine reaps stale assemblies.
 */
class Reassembler(
    scope: CoroutineScope = CoroutineScope(Dispatchers.Default),
    private val timeoutMs: Long = 10_000L,
) {
    private val pending = ConcurrentHashMap<String, Assembly>()

    init {
        scope.launch {
            while (isActive) {
                delay(5_000)
                reap()
            }
        }
    }

    /**
     * Adds a frame. Returns the reassembled data if all fragments are received and CRC is valid,
     * or null if more fragments are needed.
     * Throws [IllegalStateException] on CRC mismatch.
     */
    fun addFrame(frame: Frame): ByteArray? {
        val key = frame.packetId.joinToString("") { "%02x".format(it) }
        val count = frame.fragmentCount.toInt() and 0xFF
        val idx = frame.fragmentIndex.toInt() and 0xFF

        val assembly = pending.getOrPut(key) {
            Assembly(count = count, crc = frame.payloadCrc32)
        }

        synchronized(assembly) {
            assembly.lastSeen = System.currentTimeMillis()
            if (assembly.frags.containsKey(idx)) {
                return null // duplicate fragment
            }
            assembly.frags[idx] = frame.data

            if (assembly.frags.size < assembly.count) {
                return null // still waiting
            }

            // Reassemble
            val totalSize = (0 until assembly.count).sumOf { assembly.frags[it]!!.size }
            val buf = ByteArray(totalSize)
            var offset = 0
            for (i in 0 until assembly.count) {
                val chunk = assembly.frags[i] ?: return null
                chunk.copyInto(buf, offset)
                offset += chunk.size
            }

            pending.remove(key)

            val actualCrc = CRC32().also { it.update(buf) }.value.toInt()
            check(actualCrc == assembly.crc) {
                "CRC mismatch on packet $key: expected 0x${assembly.crc.toString(16)} got 0x${actualCrc.toString(16)}"
            }
            return buf
        }
    }

    private fun reap() {
        val now = System.currentTimeMillis()
        val stale = pending.entries.filter { now - it.value.lastSeen > timeoutMs }
        stale.forEach { pending.remove(it.key) }
    }
}

File: core/fragment.go
package core

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"hash/crc32"
	"sync"
	"time"
)

type Frame struct {
	Version       uint8
	TransferID    TransferID
	FragmentIndex uint8
	FragmentCount uint8
	PayloadCRC32  uint32
	Data          []byte
}

const (
	FrameHeaderSize = 23
	MaxFrameSize    = 200
	MaxFrameData    = MaxFrameSize - FrameHeaderSize
)

func (f Frame) Encode() []byte {
	buf := make([]byte, FrameHeaderSize+len(f.Data))
	buf[0] = f.Version
	copy(buf[1:17], f.TransferID[:])
	buf[17] = f.FragmentIndex
	buf[18] = f.FragmentCount
	binary.BigEndian.PutUint32(buf[19:23], f.PayloadCRC32)
	copy(buf[23:], f.Data)
	return buf
}

func DecodeFrame(raw []byte) (Frame, error) {
	if len(raw) < FrameHeaderSize {
		return Frame{}, fmt.Errorf("frame too short: %d bytes", len(raw))
	}
	var f Frame
	f.Version = raw[0]
	copy(f.TransferID[:], raw[1:17])
	f.FragmentIndex = raw[17]
	f.FragmentCount = raw[18]
	f.PayloadCRC32 = binary.BigEndian.Uint32(raw[19:23])
	f.Data = append([]byte(nil), raw[23:]...)
	if f.Version != FrameVersion {
		return Frame{}, fmt.Errorf("invalid frame version: %d", f.Version)
	}
	if f.FragmentCount == 0 || f.FragmentIndex >= f.FragmentCount {
		return Frame{}, fmt.Errorf("invalid fragment index/count: %d/%d", f.FragmentIndex, f.FragmentCount)
	}
	return f, nil
}

func FragmentDatagram(datagramBytes []byte, maxFrameSize int, transferID TransferID) []Frame {
	maxData := maxFrameSize - FrameHeaderSize
	if maxData < 1 {
		maxData = 1
	}
	crc := crc32.ChecksumIEEE(datagramBytes)
	var frames []Frame
	for i := 0; i < len(datagramBytes) || (len(datagramBytes) == 0 && i == 0); i += maxData {
		end := i + maxData
		if end > len(datagramBytes) {
			end = len(datagramBytes)
		}
		chunk := append([]byte(nil), datagramBytes[i:end]...)
		frames = append(frames, Frame{Version: FrameVersion, TransferID: transferID, FragmentIndex: uint8(len(frames)), PayloadCRC32: crc, Data: chunk})
		if len(datagramBytes) == 0 {
			break
		}
	}
	count := uint8(len(frames))
	for i := range frames {
		frames[i].FragmentCount = count
	}
	return frames
}

func FragmentDatagramNew(datagramBytes []byte, maxFrameSize int) []Frame {
	return FragmentDatagram(datagramBytes, maxFrameSize, NewTransferID())
}

type Reassembler struct {
	mu      sync.Mutex
	pending map[reassemblyKey]*assembly
	timeout time.Duration
}

type reassemblyKey struct {
	Peer       string
	TransferID TransferID
}

type assembly struct {
	frags    map[uint8][]byte
	count    uint8
	crc      uint32
	lastSeen time.Time
}

func NewReassembler() *Reassembler {
	r := &Reassembler{
		pending: make(map[reassemblyKey]*assembly),
		timeout: 10 * time.Second,
	}
	go r.reap()
	return r
}

func (r *Reassembler) AddFrame(peerLinkID string, f Frame) ([]byte, bool, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	k := reassemblyKey{Peer: peerLinkID, TransferID: f.TransferID}
	a, ok := r.pending[k]
	if !ok {
		a = &assembly{frags: make(map[uint8][]byte), count: f.FragmentCount, crc: f.PayloadCRC32}
		r.pending[k] = a
	}
	if a.count != f.FragmentCount || a.crc != f.PayloadCRC32 {
		delete(r.pending, k)
		return nil, false, fmt.Errorf("fragment metadata changed")
	}
	a.lastSeen = time.Now()
	if _, dup := a.frags[f.FragmentIndex]; dup {
		return nil, false, nil
	}
	a.frags[f.FragmentIndex] = f.Data
	if uint8(len(a.frags)) < a.count {
		return nil, false, nil
	}

	var buf bytes.Buffer
	for i := uint8(0); i < a.count; i++ {
		d, ok := a.frags[i]
		if !ok {
			return nil, false, nil
		}
		buf.Write(d)
	}
	delete(r.pending, k)

	data := buf.Bytes()
	if crc32.ChecksumIEEE(data) != a.crc {
		return nil, false, fmt.Errorf("CRC mismatch on transfer %x", f.TransferID)
	}
	return data, true, nil
}

func (r *Reassembler) reap() {
	ticker := time.NewTicker(5 * time.Second)
	for range ticker.C {
		r.Reap()
	}
}

func (r *Reassembler) Reap() {
	r.mu.Lock()
	now := time.Now()
	for id, a := range r.pending {
		if now.Sub(a.lastSeen) > r.timeout {
			delete(r.pending, id)
		}
	}
	r.mu.Unlock()
}
